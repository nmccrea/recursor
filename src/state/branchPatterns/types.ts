import { EntityId } from "@reduxjs/toolkit"

export type Translation = Nominal<number, "Translation">
export type Scale = Nominal<number, "Scale">
export type Angle = Nominal<number, "Angle">
export type Depth = Nominal<number, "Depth">
export type Color = Nominal<string, "Color">

export interface BranchPatternNumericInputs {
  translation: Translation
  scale: Scale
  angle: Angle
  depth: Depth
}
export type BranchPatternNumericInputKey = keyof BranchPatternNumericInputs

export interface BranchPatternInputs extends BranchPatternNumericInputs {
  color: Color
}
export type BranchPatternInputKey = keyof BranchPatternInputs

export type BranchPatternId = Nominal<EntityId, "BranchPatternId">

export interface BranchPattern extends BranchPatternInputs {
  id: BranchPatternId
}
