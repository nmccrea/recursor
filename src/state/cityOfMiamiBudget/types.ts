export enum Status {
  Idle = "idle",
  Pending = "pending",
  Fulfilled = "fulfilled",
}

export interface CityOfMiamiBudgetState {
  status: Status
  data?: object
}
