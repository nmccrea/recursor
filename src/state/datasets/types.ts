export type Dataset = object

export type DatasetId = string

export interface DatasetIndex {
  [dataset: string]: DatasetState
}

export interface Datasets {
  index: DatasetIndex
}

export type DatasetState = {
  status: Status
  data?: Dataset
}

export interface FetchResult {
  datasetId: DatasetId
  data: Dataset
}

export enum Status {
  Idle = "idle",
  Pending = "pending",
  Fulfilled = "fulfilled",
}
