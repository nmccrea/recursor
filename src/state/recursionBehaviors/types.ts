export type Translation = Nominal<number, "Translation">
export type Scale = Nominal<number, "Scale">
export type Angle = Nominal<number, "Angle">
export type Depth = Nominal<number, "Depth">
export type Color = Nominal<string, "Color">

export interface Transformation {
  translation: Translation
  scale: Scale
  angle: Angle
}

export type RecursionBehaviorId = Nominal<string, "RecursionBehaviorId">

export interface RecursionBehavior extends Transformation {
  id: RecursionBehaviorId
  depth: Depth
  color: Color
}
