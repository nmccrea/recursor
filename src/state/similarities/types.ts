import { EntityId } from "@reduxjs/toolkit"

export type Translation = Nominal<number, "Translation">
export type Scale = Nominal<number, "Scale">
export type Angle = Nominal<number, "Angle">
export type Depth = Nominal<number, "Depth">
export type Color = Nominal<string, "Color">

export interface SimilarityNumericInputs {
  translation: Translation
  scale: Scale
  angle: Angle
  depth: Depth
}
export type SimilarityNumericInputKey = keyof SimilarityNumericInputs

export interface SimilarityInputs extends SimilarityNumericInputs {
  color: Color
}
export type SimilarityInputKey = keyof SimilarityInputs

export type SimilarityId = Nominal<EntityId, "SimilarityId">

export interface Similarity extends SimilarityInputs {
  id: SimilarityId
}
