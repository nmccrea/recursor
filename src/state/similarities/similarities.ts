import { createEntityAdapter } from "@reduxjs/toolkit"
import { Similarity } from "./types"

const entityAdapter = createEntityAdapter<Similarity>()
export default entityAdapter
export type Similarities = ReturnType<typeof entityAdapter.getInitialState>
