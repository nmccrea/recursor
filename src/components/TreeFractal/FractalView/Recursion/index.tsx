import React from "react"
import styled from "styled-components"
import { useSelector } from "react-redux"
import { selectAll } from "../../../../state/similarities/selectors"
import {
  Similarity,
  Translation,
  Scale,
  Angle,
  Depth,
  Color,
} from "../../../../state/similarities/types"

interface RecursionGeometryProps {
  $translation: Translation
  $scale: Scale
  $angle: Angle
  $color: Color
}

// Use inline styles via `attrs()` for transform to avoid spamming new `styled-components` styles.
// See https://github.com/styled-components/styled-components/issues/1212#issuecomment-461134678
const RecursionGeometry = styled.div.attrs<RecursionGeometryProps>(
  ({ $translation, $scale, $angle, $color }) => ({
    style: {
      backgroundColor: $color,
      transform: `translateY(-${
        100 * $translation
      }%) scale(${$scale}) rotate(${$angle}rad)`,
    },
  })
)<RecursionGeometryProps>`
  transform-origin: center bottom;
  position: absolute;
  top: 0px;
  bottom: 0px;
  left: 0px;
  right: 0px;
`

interface RecursionProps extends Similarity {
  currentDepth: Depth
}

const Recursion = ({
  id,
  depth,
  currentDepth,
  translation,
  scale,
  angle,
  color,
}: RecursionProps) => {
  if (currentDepth >= depth) return null

  const similarities = useSelector(selectAll)
  return (
    <RecursionGeometry
      $translation={translation}
      $scale={scale}
      $angle={angle}
      $color={color}
      className={`similarity-${id} depth-${currentDepth + 1}`}
    >
      {similarities.map((similarity) => (
        <Recursion
          {...similarity}
          currentDepth={currentDepth + 1}
          key={similarity.id}
        />
      ))}
    </RecursionGeometry>
  )
}

export default Recursion
