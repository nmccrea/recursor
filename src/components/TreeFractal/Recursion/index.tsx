import React from "react"
import { useSelector } from "react-redux"
import { selectAll } from "../../../state/similarities/selectors"
import { Similarity, Depth } from "../../../state/similarities/types"

/**
 * Generates a CSS object with styles dictated by the given similarity.
 */
const styleFor = ({ translation, scale, angle, color }: Similarity) => ({
  backgroundColor: color,
  transformOrigin: "center bottom",
  transform: [
    `translateY(-${100 * translation}%)`,
    `scale(${scale})`,
    `rotate(${angle}rad)`,
  ].join(" "),
})

interface RecursionProps {
  similarity: Similarity
  currentDepth: Depth
}

const Recursion = ({ similarity, currentDepth }: RecursionProps) => {
  if (currentDepth >= similarity.depth) return null
  const similarities = useSelector(selectAll)
  return (
    <div
      className={`tree ${similarity.id} depth-${currentDepth + 1}`}
      style={{
        position: "absolute",
        inset: 0,
        ...styleFor(similarity),
      }}
    >
      {similarities.map((similarity) => (
        <Recursion
          similarity={similarity}
          currentDepth={currentDepth + 1}
          key={`${similarity.id}`}
        />
      ))}
    </div>
  )
}

export default Recursion
