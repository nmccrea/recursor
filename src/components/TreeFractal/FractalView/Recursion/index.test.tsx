import React from "react"
import "jest-styled-components"
import { Provider } from "react-redux"
import { render } from "@testing-library/react"
import { createStore, RootState } from "state/store"
import { Similarity } from "state/similarities/types"
import Recursion from "."

const SIMILARITY_A: Similarity = {
  id: "test/a",
  translation: 0.994,
  scale: 0.543,
  angle: -3.012 * Math.PI,
  depth: 2,
  color: "green",
}

const SIMILARITY_B: Similarity = {
  id: "test/b",
  translation: 1.042,
  scale: 0.452,
  angle: 1.902 * Math.PI,
  depth: 1,
  color: "red",
}

const TEST_STATE: RootState = {
  similarities: {
    ids: [SIMILARITY_A.id, SIMILARITY_B.id],
    entities: {
      [SIMILARITY_A.id]: SIMILARITY_A,
      [SIMILARITY_B.id]: SIMILARITY_B,
    },
  },
}

describe("`<Recursion />`", () => {
  it("correctly renders the tree with the appropriate depth", () => {
    const { container } = render(
      <Provider store={createStore(TEST_STATE)}>
        <Recursion {...SIMILARITY_A} currentDepth={0} />
      </Provider>
    )

    expect(container).toMatchSnapshot()
  })

  it("renders nothing when the `currentDepth` is equal to the given similarity's depth", () => {
    const { container } = render(
      <Provider store={createStore(TEST_STATE)}>
        <Recursion {...SIMILARITY_A} currentDepth={2} />
      </Provider>
    )

    expect(container).toMatchSnapshot()
  })
})
