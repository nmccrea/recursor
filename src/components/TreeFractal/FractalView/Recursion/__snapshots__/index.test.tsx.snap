// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`<Recursion />\` correctly renders the tree with the appropriate depth 1`] = `
<div>
  <div
    class="tree test/a depth-1"
    style="position: absolute; inset: 0; background-color: green; transform-origin: center bottom; transform: translateY(-99.4%) scale(0.543) rotate(-9.462477072612456rad);"
  >
    <div
      class="tree test/a depth-2"
      style="position: absolute; inset: 0; background-color: green; transform-origin: center bottom; transform: translateY(-99.4%) scale(0.543) rotate(-9.462477072612456rad);"
    >
      <div
        class="tree test/a depth-3"
        style="position: absolute; inset: 0; background-color: green; transform-origin: center bottom; transform: translateY(-99.4%) scale(0.543) rotate(-9.462477072612456rad);"
      />
    </div>
    <div
      class="tree test/b depth-2"
      style="position: absolute; inset: 0; background-color: red; transform-origin: center bottom; transform: translateY(-104.2%) scale(0.452) rotate(5.975309227127786rad);"
    >
      <div
        class="tree test/a depth-3"
        style="position: absolute; inset: 0; background-color: green; transform-origin: center bottom; transform: translateY(-99.4%) scale(0.543) rotate(-9.462477072612456rad);"
      />
    </div>
  </div>
</div>
`;

exports[`\`<Recursion />\` renders nothing when the \`currentDepth\` is equal to the given similarity's depth 1`] = `<div />`;
